menelaus/injection/class_manipulation.py

```diff
@@ -94,3 +96,119 @@ def class_join(data, target_col, class_1, class_2, new_class, from_index, to_ind
         class_idx = class_idx[(class_idx < to_index) & (class_idx >= from_index)]
         ret[class_idx, target_col] = new_class
     return ret
+
+# endregion
+
+
+# region - LTF-inspired class manipulation functions
+
+def class_probability_shift(data, target_col, from_index, to_index, classes, p_classes):
```

Comment:
Here’s how I’m understanding the purpose of this function: 
Probability_shift receives a float as input and sets the proportion of the specified label to be that float, while other classes are distributed uniform. If multiple labels are specified, it sets the proportion of each of those classes to the specified float, with the rest uniform. 
-
This method implements the tweak one shift from LTF. If wanting to implement minority-class shift from LTF, you would simply pass a minority class label and desired proportion. 

We might need to add in an error catch statement to make sure p_classes*the number of classes input to ‘classes’ does not exceed 1?

Reply:
Yeah, my thinking matches the above^

I'll add the check, I was thinking we may need something or another given the different combinations in which this method can be called

Reply:
I think the code is treating the parameter a little differently than India's expecting, maybe.

India's description, put another way - if you had five classes, two of which were specified to be sampled with probability `p_classes`, then the remaining 3 categories would need to be resampled, uniform with probability `(1 - 2*p_classes)/3`.

If I'm reading it right, the code's behavior would be to have new probability of `p_classes/2` for each of the classes you're upsampling (L146), and then `(1 - p_classes)/3` (L149).

With the latter, you don't need a check. With India's parameterization, you would need something to make sure that `p_classes * len(classes) <= 1` (or something to scale the user's input down to a valid probability vector, which seems a little unclean).

Reply:
I'm not sure if I'd find the `p_classes` parameter being reduced by the number of `classes` intuitive as a user.

If this function were more general, with list or dict parameters, could we re-use it? e.g.
```
class_probability_shift(..., foo, ...)
  """
  foo (dict) - dictionary of the form {class, probability} that defines new sampling parameters. 
                   Undefined classes are resampled uniformly.
  """
  ...

class_dirichlet_shift(...)
  """
  alpha - parameter for Dirichlet distribution ...
  """
  new_vec = dirichlet_distribution(alpha)
  class_probability_shift(..., foo=new_vec, ...)
```

Would something like that work?

Reply:
It'll be helpful to think through an example to all be on the same page. 

If we have 100 samples and there are 5 possible labels (0-4). The user wants to increase labels 0, 1, and 2 to appear with probability = 0.3, and labels 4, 5 to be uniform (but not disappear). The user passes in p_classes = 0.3 and classes = [1,2,3]. @class_probability_shift should assign:
- 30 samples for label 0 (0.3*total samples)
- 30 for label 1: 0.3*total samples
- 30 for label 2: 0.3*total samples
- 5 for label 4:  (1-0.3*3)*total samples
- 5 for label 5:  (1-0.3*3)*total samples

We would need an error statement to make sure p_classes*len(classes) <= 1. For example, passing in 4 class labels and p_classes = 0.3 would result in a 'error'. 

Or do we want this to function in some other way?

Reply:
Then for the dirichlet, that function allows a user to specify different proportions to a dirichlet distribution for different classes. 

For the same example, a user could want {0:0.2, 1:0.3, 2:0.4, 3:0.05, 4:0.05}. So we would need first to check to make sure 0.2 + 0.3 + 0.4 + 0.05 + 0.05 <= 1. We pass alpha = [0.2, 0.3, 0.4, 0.05, 0.05] to the np.dirichlet method and it returns a proportion to sample those classes from the total number of observations in the data, according to the dirichlet distribution. (I believe the np.dirichlet method will not work if the sum(alpha) > 1). 

Following off of your second comment @tms-bananaquit, I don't think we could pass that into @class_proportion_shift since new_vec would contain an array of different proportions, not one proportion as expected by the p_classes parameter.